### PR TITLE
DUPLO-20918: [Avialabilty Zone] RDS primary and replica to support AZ zone

### PR DESCRIPTION
**Pull request by pravin-duplocloud**
### **User description**
DUPLO-20918: [Avialabilty Zone] RDS primary and replica to support AZ zone

## Overview

... (don't forget to put your clickup ticket ID in the title) ...

## Summary of changes

This PR does the following:

- availability zone for aurora writer
- availability zone for rds primary

## Testing performed

- [ ] Using unit tests
- [X] Manually, on my local system
- [X] Manually, on a remote test system

## Describe any breaking changes

- ...

___

### **PR Type**
enhancement, documentation

___

### **Description**
- Added a new `availability_zone` field to the RDS instance schema, allowing users to specify a valid Availability Zone for the RDS primary or Aurora writer instances.
- Implemented validation to ensure that `multi_az` and `availability_zone` cannot be set simultaneously.
- Updated documentation to include details and examples for the new `availability_zone` field.
- Updated the changelog to reflect the enhancements made in this PR.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>resource_duplo_rds_instance.go</strong><dd><code>Add and validate `availability_zone` field in RDS schema</code>&nbsp; </dd></summary>
<hr>

duplocloud/resource_duplo_rds_instance.go

<li>Added <code>availability_zone</code> field to RDS instance schema.<br> <li> Implemented validation to prevent <code>multi_az</code> and <code>availability_zone</code> from <br>being set together.<br> <li> Updated state functions to handle the new <code>availability_zone</code> field.<br>

</details>

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/648/files#diff-a8bf9b56a3e9b39899a5d37d7608110678908754e0848fab2d0320653ea65f71">+13/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>CHANGELOG.md</strong><dd><code>Update changelog with `availability_zone` enhancement</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

CHANGELOG.md

<li>Documented the addition of <code>availability_zone</code> field to RDS instance <br>schema.<br> <li> Included validation details for the new field.<br>

</details>

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/648/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4ed">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>rds_instance.md</strong><dd><code>Document `availability_zone` field for RDS instances</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

docs/resources/rds_instance.md

<li>Added documentation for the <code>availability_zone</code> field in RDS instances.<br> <li> Provided example usage of the new field.<br>

</details>

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/648/files#diff-2be785952795586c48f30db78e5f3684f72a430f66155dd8ea0d85f0607eedb7">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>resource.tf</strong><dd><code>Add example for `availability_zone` in RDS resource</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

examples/resources/duplocloud_rds_instance/resource.tf

- Added `availability_zone` example to RDS instance resource.

</details>

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/648/files#diff-4461dc184c959908eb1db51855cf42bfafd849c340d3632e8da923f5a5a6240d">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: docs/resources/rds_instance.md [hunk 1]
```diff
@@ -37,6 +37,10 @@ resource "duplocloud_rds_instance" "mydb" {
 
   encrypt_storage         = true
   backup_retention_period = 1
+  # Specify a valid Availability Zone for
+  # the RDS primary instance (when Multi-AZ is disabled)
+  # or for the Aurora writer instance.
+  availability_zone = true
```
Comment (zafarabbas) [inline docs/resources/rds_instance.md:43]: This is not a boolean, its an AZ right?

**Review by zafarabbas — CHANGES_REQUESTED** [review]

File: docs/resources/rds_instance.md [hunk 2]
```diff
@@ -37,6 +37,10 @@ resource "duplocloud_rds_instance" "mydb" {
 
   encrypt_storage         = true
   backup_retention_period = 1
+  # Specify a valid Availability Zone for
+  # the RDS primary instance (when Multi-AZ is disabled)
+  # or for the Aurora writer instance.
+  availability_zone = us-west-2a
```
Comment (zafarabbas) [inline docs/resources/rds_instance.md:43]: we can remove the comment from the example, the field is already documented below in detail, this clutters the example code unnecessarily

**Review by zafarabbas — CHANGES_REQUESTED** [review]

File: docs/resources/rds_instance.md (hunk 2 above)
Comment (pravin-duplocloud) [reply docs/resources/rds_instance.md]: done

File: docs/resources/rds_instance.md [hunk 3]
```diff
@@ -37,10 +37,7 @@ resource "duplocloud_rds_instance" "mydb" {
 
   encrypt_storage         = true
   backup_retention_period = 1
-  # Specify a valid Availability Zone for
-  # the RDS primary instance (when Multi-AZ is disabled)
-  # or for the Aurora writer instance.
-  availability_zone = us-west-2a
+  availability_zone       = us-west-2a
```
Comment (zafarabbas) [inline docs/resources/rds_instance.md:40]: is zone supposed to be a string like “us-west-2a”?

File: duplocloud/resource_duplo_rds_instance.go [hunk 4]
```diff
@@ -289,6 +289,14 @@ func rdsInstanceSchema() map[string]*schema.Schema {
...
 		},
+		"availability_zone": {
+			Description: "Specify a valid Availability Zone for the RDS primary instance" +
+				" (when Multi-AZ is disabled) or for the Aurora writer instance." +
+				" e.g. us-west-2a",
+			Type:     schema.TypeString,
+			Computed: true,
+			Optional: true,
+		},
 	}
```
Comment (Duplo-nikv) [inline duplocloud/resource_duplo_rds_instance.go:300]: I think availability_zone should be set [us-west-2a,us-west-2b]

**Review by Duplo-nikv — CHANGES_REQUESTED** [review]

File: docs/resources/rds_instance.md (hunk 1 above)
Comment (pravin-duplocloud) [reply docs/resources/rds_instance.md]: done

File: docs/resources/rds_instance.md (hunk 2 above)
Comment (pravin-duplocloud) [reply docs/resources/rds_instance.md]: done

File: duplocloud/resource_duplo_rds_instance.go (hunk 4 above)
Comment (pravin-duplocloud) [reply duplocloud/resource_duplo_rds_instance.go:300]: this is for rds primary and aurora write specifically, not considering  cluster.
- customer want to control master in default zone and fialover replica in another zone.

**Review by zafarabbas — DISMISSED** [review]

File: duplocloud/resource_duplo_rds_instance.go (hunk 4 above)
Comment (Duplo-nikv) [reply duplocloud/resource_duplo_rds_instance.go:300]: ok closing based on the discussion

**Review by Duplo-nikv — APPROVED** [review]
LGTM

**Review by tahir-duplo — APPROVED** [review]
LGTM

File: docs/resources/rds_instance.md (hunk 3 above)
Comment (pravin-duplocloud) [reply docs/resources/rds_instance.md]: done

**Pull request merged**